### PR TITLE
Freecall Support in Daemon - 

### PR DESCRIPTION
**Pull request by anandrgitnirman**
#439 
More test cases are in progress .. but a working version 
Will break this in to smaller PRs for final review

### CODE REVIEW

File: escrow/free_call.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		ServiceId: payment.ServiceId, GroupID: blockchain.BytesToBase64(groupId[:])}
+	freeCallUserData, ok, err := h.FreeCallUserUsage(userKey)
+	if err != nil {
+		return nil, NewPaymentError(Internal, "payment freeCallUserData error:"+err.Error())
+	}
+	if !ok {
+		log.Warn("Payment freeCallUserData not found")
+		return nil, NewPaymentError(Unauthenticated, "payment freeCallUserData \"%v\" not found", userKey)
+	}
+	if err != nil {
```
Comment (raamb) [inline escrow/free_call.go]: Do we wait to get a mutex or retry if we dont get the first time?

File: escrow/free_call.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, NewPaymentError(Internal, "cannot get mutex for user: %v", userKey)
+	}
+	if !ok {
+		return nil, NewPaymentError(FailedPrecondition, "another transaction on this user: %v is in progress", userKey)
+	}
+	defer func(lock Lock) {
+		if err != nil {
+			e := lock.Unlock()
+			if e != nil {
+				log.WithError(e).WithField("userKey", userKey).WithField("err", err).Error("Transaction is cancelled because of err, but freeCallUserData cannot be unlocked. All other transactions on this freeCallUserData will be blocked until unlock. Please unlock freeCallUserData manually.")
```
Comment (raamb) [inline escrow/free_call.go]: Should we send a notification in this case? if this code executes the user is essentially completely blocked so we should make it easy for the service developer to remedy this

File: escrow/free_call.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+func (transaction *freeCallTransaction) Commit() error {
+	defer func(payment *freeCallTransaction) {
+		err := payment.lock.Unlock()
+		if err != nil {
+			log.WithError(err).WithField("transaction", payment).
+				Error("free call user cannot be unlocked because of error." +
+					" All other transactions on this channel will be blocked until unlock." +
+					" Please unlock user for free calls manually.")
+		} else {
```
Comment (raamb) [inline escrow/free_call.go]: Do we need this?

File: escrow/free_call_api.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"math/big"
+
+	"github.com/singnet/snet-daemon/blockchain"
+)
+
+func FreeCallID(userAddress *common.Address, serviceId string) string {
+	return fmt.Sprintf("%v/%v", userAddress, serviceId)
```
Comment (raamb) [inline escrow/free_call_api.go]: We should include orgID as well right?

File: escrow/free_call_api.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func (key *FreeCallUserKey) String() string {
+	return fmt.Sprintf("{ID:%v/%v/%v/%v}", key.UserId, key.OrganizationId,
+		key.ServiceId, key.GroupID)
+}
+
+type FreeCallUserData struct {
+	UserId        string
+	ServiceId     string
+	OrgId         string
+	GroupID       [32]byte
```
Comment (ksridharbabuus) [inline escrow/free_call_api.go]: Keep GroupID as consistent with ServiceId and OrgId.
Keep the OrgId as consistent naming convention where some places it is mentioned as OrganizationId.

File: escrow/free_call.go (hunk 2 above)
Comment (anandrgitnirman) [reply escrow/free_call.go]: we can do this for the channel locks too

**Pull request closed without merging**